File: javascript/robot-name/robot-name.js
function Robot() {
  function makeRandomName() {
    var LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var NUMBERS = "0123456789";

    function getChars(charSet, quantity) {
      var chars = ''
      for ( var i = 0; i < quantity; i++ )
        chars += charSet.charAt(Math.floor(Math.random() * charSet.length));
      return chars;
    }

    return getChars(LETTERS, 2) + getChars(NUMBERS, 3);
  }

  function nameExists(name) {
    return Robot.currentNames.indexOf(name) >= 0;
  }

  function replaceNamesInList(nameToRemove, nameToAdd) {
    var indexOfOldName = Robot.currentNames.indexOf(nameToRemove);

    if (indexOfOldName > -1) {
      Robot.currentNames.splice(indexOfOldName, 1);
    }
    Robot.currentNames.push(nameToAdd);
  }

  this.reset = function() {
    var newName = makeRandomName();

    if (nameExists(newName)) {
      this.reset();
      return;
    }

    replaceNamesInList(this.name, newName);
    this.name = newName;
  }

  this.reset();
}

Robot.currentNames = [];

module.exports = Robot;

File: ruby/robot-name/robot_name.rb
class Robot
  attr_reader :name
  @@all_names = []

  def initialize
    reset
  end

  def reset
    newName = create_name

    if (@@all_names.include? newName)
      reset
      return
    end

    @@all_names.map! { |n| n == @name ? newName : n }
    @name = newName
  end

  private

  def create_name
    (('A'..'Z').to_a.sample(2) + (0..9).to_a.sample(3)).join
  end
end
